### ISSUE TITLE
Linq .Any() without predicate - error with default TypeNameHandling

### ISSUE THREAD

**Issue by robert-skarzycki** [JasperFx/marten#1192]
Having that model (https://gist.github.com/robert-skarzycki/c0c26f32e0000c17faabc6c5debde55c) and data inserting code (https://gist.github.com/robert-skarzycki/aa92e11a27fdb33ea8f89e59ea90329d), this query fails:
```
var res = session
.Query<Root>()
.Where(r => r.ChildsLevel1.Any());
```
with following exception:
```
'Marten Command Failure:
select count(*) as number from public.mt_doc_root as d where JSONB_ARRAY_LENGTH(COALESCE(case when data ->> 'ChildsLevel1' is not null then data -> 'ChildsLevel1' else '[]' end)) > 0

22023: cannot get array length of a non-array'
```

I've found that reconfiguring JSON serializer to prevent storing types in Postgres (by setting `TypeNameHandling = TypeNameHandling.None`) fixes that. I guess it should either work, or the exception message should be more descriptive.

P.S. I'm using Marten `3.4.0` on .NET Core 2.2.

**Comment by oskardudycz** [comment]
Thank you for creating this issue. As I stated in [my comment in other issue](https://github.com/JasperFx/marten/pull/1107#issuecomment-464845482) I'll try to work on that asap, but I can't put specific date where the fix will be delivered.

**Comment by jeremydmiller** [comment]
@oskardudycz I'll take this one while I'm into all the Linq stuff

**Comment by mysticmind** [comment]
The problem is primarily around serialized document being stored with `$type` and `$value` rather than as an array for `ICollection`.

I have written the below Custom JSON converter for handling `ICollection<T>`,`IReadOnlyCollection<T>`, `IEnumerable<T>` and also any of these types assigned with an T[] object i.e. array results in a `$type` and $value as well. 

Note that for all above types, JsonNetSerializer will store the `$type` and `$value` if the custom converter is not used.

```
public class CustomConverter<T> : JsonConverter
    {
        private readonly List<Type> _types = new List<Type>
        {
            typeof(ICollection<T>)
            typeof(IReadOnlyCollection<T>),
            typeof(IEnumerable<T>),
            typeof(T[])
        };
        
        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
        {
            var o = value as IEnumerable<T>;
            serializer.Serialize(writer, o?.ToArray());
        }

        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {
            return serializer.Deserialize<List<T>>(reader);
        }

        public override bool CanConvert(Type objectType)
        {
            return _types.Contains(objectType);
        }
    }
```

Add the JsonConverter attribute as below to `ICollection` properties as an example:
```
public class Root
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
        [JsonConverter(typeof(CustomConverter<string>))] 
        public ICollection<ChildLevel1> ChildsLevel1 { get; set; }
    }

    public class ChildLevel1
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
        [JsonConverter(typeof(CustomConverter<string>))]
        public ICollection<ChildLevel2> ChildsLevel2 { get; set; }
    }

    public class ChildLevel2
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
}
```
Above fix also applies for the tests in PR #1107, I have tested it and all tests are green. Custom json converter is better than using a blanket setting of `Newtonsoft.Json.TypeNameHandling.None`,

**Comment by oskardudycz** [comment]
Cool @mysticmind ! Maybe we could think about plugin that as optional choice for collection handling? Or use it as default in `4.0`?

**Comment by mysticmind** [comment]
@oskardudycz I think we can include the JsonConverter as a utility class as part of Marten. Note that the user will need to add the attribute to the class properties accordingly.

**Comment by jeremydmiller** [comment]
Closed by #1107
